packages/flutter_tools/lib/src/http_host_availability_validator.dart

```diff
@@ -0,0 +1,148 @@
+// Copyright 2014 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+// @dart = 2.9
+
+import 'dart:async';
+
+import 'package:meta/meta.dart';
+
+import 'base/io.dart';
+import 'base/platform.dart';
+import 'doctor_validator.dart';
+import 'features.dart';
+
+// The environment variables used to override some URLs
+const String kPubHostedUrl = 'PUB_HOSTED_URL';
+const String kCloudUrl = 'FLUTTER_STORAGE_BASE_URL';
+
+/// Validation class that checks if all the given URLs are reachable
+class HttpHostAvailabilityValidator extends DoctorValidator {
+  HttpHostAvailabilityValidator({
+    @required Platform platform,
+    @required FeatureFlags featureFlags,
+    @required HttpClient httpClient,
+  })   : _platform = platform,
+        _featureFlags = featureFlags,
+        _httpClient = httpClient,
+        super('HTTP host availability');
+
+  final Platform _platform;
+  final FeatureFlags _featureFlags;
+  final HttpClient _httpClient;
+
+  @override
+  String get slowWarning => 'HTTP host availability check is taking a long time';
+
+  /// Returns a list of URLs to check availability, different for different platforms
+  List<String> get _allRequiredHosts {
+    /// Hosts used by flutter
+    final List<String> requiredHostUrls = <String>[
+      if (_featureFlags.isAndroidEnabled)
+        'https://maven.google.com/',
+      if (_featureFlags.isIOSEnabled)
+        'https://cocoapods.org/',
+      if (_platform.environment.containsKey(kCloudUrl))
+        _platform.environment[kCloudUrl]
+      else
+        'https://cloud.google.com/',
+      if (_platform.environment.containsKey(kPubHostedUrl))
+        _platform.environment[kPubHostedUrl]
+      else
+        'https://pub.dev/',
+    ];
+
+    return requiredHostUrls;
+  }
+
+  /// Make an HTTP HEAD request to the given URL. If there is no exception,
+  /// the host is available. If there is an exception, return a failed result.
+  Future<_HttpHostAvailabilityResult> _checkHostAvailability(
+    String hostUrl,
+  ) async {
+    try {
+      // Make the HEAD request
+      final HttpClientRequest headReq = await _httpClient.headUrl(Uri.parse(hostUrl));
+      await headReq.close().timeout(const Duration(seconds: 10));
```

Comment:
Try making the timeout configurable and then set it to a fairly low time for testing?